spec/support/metric_helper.rb

```diff
@@ -1,46 +1,54 @@
-      def metric_targets(expected_targets)
-        expected_targets.flat_map do |t|
-          # Storage is hourly only
-          # Non-storage historical is expecting 7 days back, plus partial day = 8
-          t.kind_of?(Storage) ? [[t, "hourly"]] : [[t, "realtime"]] + [[t, "historical"]] * 8
-        end
-      end
-
-      # @return [Array<Array<Object, String>>] List of object and interval names in miq queue
-      def queue_intervals(items = MiqQueue.where(:method_name => %w[perf_capture_hourly perf_capture_realtime perf_capture_historical]))
-        items.map do |q|
-          interval_name = q.method_name.sub("perf_capture_", "")
-          [Object.const_get(q.class_name).find(q.instance_id), interval_name]
-        end
-      end
-
       # method_name => {target => [timing1, timing2] }
       # for each capture type, what objects are submitted and what are their time frames
       # @return [Hash{String => Hash{Object => Array<Array>}} ]
       def queue_timings(items = MiqQueue.where(:method_name => %w[perf_capture_hourly perf_capture_realtime perf_capture_historical]))
         messages = {}
         items.each do |q|
-          obj = q.instance_id ? Object.const_get(q.class_name).find(q.instance_id) : q.class_name.constantize
+          klass = q.class_name.constantize.base_class.name
+          # exploding looks into the third argument which contains the list of object ids
+          ids = q.args[2]
+          ids = [q.instance_id] if ids.blank?
+          objs = ids.map { |id| "#{klass}:#{id}" }.sort
 
-          interval_name = q.method_name.sub("perf_capture_", "")
+          objs.each do |obj|
+            interval_name = q.method_name.sub("perf_capture_", "")
 
-          messages[interval_name] ||= {}
-          (messages[interval_name][obj] ||= []) << q.args
+            messages[interval_name] ||= {}
+            (messages[interval_name][obj] ||= []) << q.args[0..1]
+          end
         end
-        messages["historical"]&.transform_values!(&:sort!)
+        # just focus on the ranges being the same and ignore the date splitups
+        messages["historical"]&.transform_values! { |v| combine_consecutive(v) }
 
         messages
       end
 
-      # sorry, stole from the code - not really testing
+      # to make test failures easier to read, use parent_class:id
+      def queue_object(object)
+        "#{object.class.base_class.name}:#{object.id}"
+      end
+
+      def combine_consecutive(array)
+        x = array.sort!.shift
+        array.each_with_object([]) do |i, ac|
+          # we have 2 options:
+          #   x=1-1-2000T0..23:59, i=1-2-2000T0..23:59
+          #   x=1-1-2000T0..24:00, i=1-2-2000T0..24:00
+          # so they will be within a second of eachother
+          if i.first <= (x.second + 1.second) # they overlap
+            x[1] = i.second
+          else
+            ac << x
+            x = i
+          end
+        end << x
```

Comment:
Using `x.second` (i.e. the second parameter) and `1.second` (i.e. 1 time unit second) in the same line is SUPER confusing.  This might be a lot more readable with named variables...something like the below. Probably helpful to name things instead of just `x` also.

Thinking something like:

```suggestion
        prev_range_start, prev_range_end = array.sort!.shift
        array.each_with_object([]) do |(range_start, range_end), ac|
          # we have 2 options:
          #   x=1-1-2000T0..23:59, i=1-2-2000T0..23:59
          #   x=1-1-2000T0..24:00, i=1-2-2000T0..24:00
          # so they will be within a second of each other
          if range_start <= (prev_range_end + 1.second) # they overlap
            prev_range_end = i.second
          else
            ac << [prev_range_start, prev_range_end]
            range_start, range_end = prev_range_start, prev_range_end
          end
        end << [prev_range_start, prev_range_end]
```

Reply:
Now that I wrote that I think there are bugs in it?  Can you write some specs demonstrating various cases?

I'm also concerned that it modifies the incoming array with `.sort!.shift`, but that is being used in a `transform_values!` call, so I'm not sure if there are any bad side effects to that.

Reply:
Separately, asking the larger question, how do we even get times that are within a second of each other.  I thought it was built by stepping day by day and the initial day is at beginning_of_day?

Reply:
sometimes we do beginning of day end of day, which is 1 second off

Reply:
ASIDE: I had wanted to use a range instead of a start/end pair, but surprisingly the code was more confusing.